SIMD: Use universal intrinsics to implement comparison functions

rafaelcfsousa opened the pull request:
This PR moves the comparison functions (eq, ne, lt, le, ge and gt) to a new dispatchable source file to make use of the universal intrinsics. This optimization benefits all architectures.

The following universal intrinsics were added in this PR:
```
npyv_b8 npyv_pack_b8_b16(npyv_b16 a, npyv_b16 b);
npyv_b8 npyv_pack_b8_b32(npyv_b32 a, npyv_b32 b, npyv_b32  c, npyv_b32 d);
npyv_b8 npyv_pack_b8_b64(npyv_b64 a, npyv_b64 b, npyv_b64  c, npyv_b64 d, 
                         npyv_b64 e, npyv_b64 f, npyv_b64  g, npyv_b64 h);
```

## Benchmark:

### X86

<details>
  <summary>CPU</summary>
  
```
CPU
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                32
On-line CPU(s) list:   0-31
Thread(s) per core:    2
Core(s) per socket:    8
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Silver 4208 CPU @ 2.10GHz
Stepping:              7
CPU MHz:               2100.128
CPU max MHz:           2100.0000
CPU min MHz:           800.0000
BogoMIPS:              4200.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              11264K
NUMA node0 CPU(s):     0-7,16-23
NUMA node1 CPU(s):     8-15,24-31
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology non
stop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cd
p_l3 invpcid_single intel_ppin intel_pt ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt cl
wb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm arat pln pts pku ospke avx512_vnni md_clear spec_ctrl intel_stibp flush_l1d arch_capabilities
```
</details>

<details>
  <summary>OS</summary>

```
CentOS Linux 7 (Core)"
gcc (Spack GCC) 9.4.0
```
</details>

#### Benchmark

<details>
  <summary>SSE</summary>

```
export NPY_DISABLE_CPU_FEATURES="SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_SKX"
        before          after
+     10.1±0.04μs      11.6±0.05μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float32'>)
+     8.43±0.03μs      9.16±0.03μs     1.09  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint16'>)
+      41.3±0.1μs         43.7±1μs     1.06  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-      18.3±0.2μs      16.9±0.09μs     0.92  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-      18.1±0.1μs      16.1±0.03μs     0.89  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
-      30.2±0.4μs       26.3±0.2μs     0.87  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-     50.9±0.09μs       44.2±0.5μs     0.87  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-     8.23±0.08μs      6.97±0.01μs     0.85  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-      8.88±0.1μs      7.50±0.04μs     0.85  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-      50.8±0.2μs       42.1±0.2μs     0.83  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-     18.1±0.02μs       14.9±0.3μs     0.82  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-     8.15±0.09μs      6.47±0.04μs     0.79  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
-     50.5±0.05μs      39.3±0.05μs     0.78  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-     10.2±0.02μs      7.80±0.03μs     0.77  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-      50.7±0.1μs       38.6±0.1μs     0.76  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-      20.3±0.2μs      14.7±0.03μs     0.72  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-      19.9±0.3μs       12.3±0.1μs     0.62  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-     19.3±0.04μs      11.7±0.07μs     0.60  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-     19.0±0.01μs      10.1±0.01μs     0.53  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-     19.1±0.02μs      9.69±0.04μs     0.51  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-     77.7±0.05μs      6.92±0.01μs     0.09  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-     82.7±0.05μs      6.79±0.02μs     0.08  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
-     77.7±0.04μs      6.38±0.04μs     0.08  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)

```
</details>

<details>
  <summary>AVX2</summary>

```
export NPY_DISABLE_CPU_FEATURES="AVX512F AVX512_SKX"
        before          after
+     4.80±0.03μs      5.29±0.02μs     1.10  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint8'>)
-     9.95±0.01μs      9.44±0.01μs     0.95  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float32'>)
-     10.0±0.04μs      9.42±0.01μs     0.94  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float32'>)
-     14.6±0.08μs       13.1±0.2μs     0.90  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-     7.84±0.09μs      6.98±0.04μs     0.89  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-     14.8±0.07μs      13.0±0.05μs     0.88  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
-     7.31±0.03μs      6.32±0.04μs     0.86  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
-      7.94±0.1μs      6.85±0.06μs     0.86  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-     7.29±0.02μs      6.28±0.08μs     0.86  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-     18.2±0.08μs      15.1±0.04μs     0.83  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
-     18.6±0.07μs      15.2±0.07μs     0.82  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-      29.3±0.2μs       23.8±0.3μs     0.81  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-     11.4±0.04μs      9.26±0.04μs     0.81  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-      31.0±0.4μs       24.7±0.1μs     0.79  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-      11.7±0.2μs      9.25±0.06μs     0.79  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-      34.6±0.8μs       27.1±0.2μs     0.78  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)
-      12.7±0.2μs      9.83±0.03μs     0.77  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-      12.7±0.2μs      9.67±0.02μs     0.76  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-     28.2±0.01μs      17.8±0.03μs     0.63  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-      28.4±0.1μs      17.7±0.03μs     0.62  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-      28.4±0.1μs      17.7±0.02μs     0.62  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-      28.3±0.1μs      17.6±0.02μs     0.62  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-     82.7±0.03μs      6.03±0.01μs     0.07  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
-     77.7±0.02μs      5.42±0.02μs     0.07  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
-      77.8±0.2μs      5.37±0.06μs     0.07  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
```
</details>

<details>
  <summary>AVX512F</summary>

```
export NPY_DISABLE_CPU_FEATURES="AVX512_SKX"
        before          after
+     18.5±0.09μs       19.5±0.2μs     1.06  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-      7.19±0.1μs       6.83±0.2μs     0.95  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-      5.02±0.1μs       4.72±0.2μs     0.94  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint8'>)
-      5.04±0.2μs       4.69±0.2μs     0.93  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint8'>)
-     4.99±0.08μs       4.63±0.1μs     0.93  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int8'>)
-     4.69±0.02μs      4.32±0.06μs     0.92  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint8'>)
-      7.39±0.1μs       6.73±0.2μs     0.91  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
-     7.85±0.06μs       7.01±0.3μs     0.89  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-     8.26±0.02μs       7.26±0.1μs     0.88  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int16'>)
-      11.6±0.2μs       10.1±0.1μs     0.87  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-      7.92±0.1μs       6.79±0.3μs     0.86  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-     12.1±0.07μs       10.2±0.1μs     0.84  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-     8.36±0.03μs       7.00±0.2μs     0.84  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint16'>)
-      12.5±0.1μs       10.0±0.2μs     0.80  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-      13.1±0.3μs       10.1±0.1μs     0.77  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-      30.4±0.6μs       22.9±0.9μs     0.75  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-      14.2±0.4μs      10.1±0.04μs     0.71  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-      14.0±0.2μs      9.94±0.09μs     0.71  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-      31.0±0.9μs       21.8±0.6μs     0.70  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-      28.4±0.1μs       19.7±0.3μs     0.70  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-      28.4±0.1μs       19.7±0.1μs     0.69  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-      28.4±0.1μs       19.6±0.1μs     0.69  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-      28.6±0.1μs       19.7±0.2μs     0.69  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-      14.8±0.1μs      10.0±0.06μs     0.68  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
-      34.3±0.8μs       21.8±0.4μs     0.63  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)
-      82.5±0.2μs      6.08±0.09μs     0.07  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
-     77.7±0.07μs       5.61±0.2μs     0.07  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-      77.9±0.5μs       5.60±0.2μs     0.07  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
```
</details>

<details>
  <summary>AVX512BW</summary>

```
unset NPY_DISABLE_CPU_FEATURES
        before          after
-     4.87±0.04μs      4.54±0.08μs     0.93  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int8'>)
-     4.92±0.02μs      4.56±0.04μs     0.93  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint8'>)
-     4.88±0.01μs      4.52±0.05μs     0.93  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int8'>)
-     4.95±0.04μs      4.54±0.03μs     0.92  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint8'>)
-     4.72±0.01μs      4.21±0.07μs     0.89  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int8'>)
-     4.73±0.02μs      4.21±0.03μs     0.89  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint8'>)
-     7.09±0.07μs      5.82±0.05μs     0.82  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
-     7.04±0.06μs      5.62±0.07μs     0.80  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-     8.26±0.03μs      6.24±0.04μs     0.76  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int16'>)
-     7.60±0.06μs      5.67±0.04μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-     9.94±0.03μs       7.38±0.4μs     0.74  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float32'>)
-     8.38±0.05μs      6.22±0.04μs     0.74  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint16'>)
-     7.64±0.07μs      5.62±0.05μs     0.74  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-     10.1±0.05μs       7.32±0.4μs     0.73  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float32'>)
-     13.3±0.04μs       8.81±0.7μs     0.66  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-     14.9±0.08μs       9.44±0.2μs     0.63  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
-      11.8±0.3μs       7.36±0.3μs     0.62  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-      11.7±0.2μs       7.27±0.4μs     0.62  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-      18.2±0.3μs       11.3±0.2μs     0.62  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
-      29.9±0.5μs       18.4±0.6μs     0.62  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-      31.6±0.6μs         19.3±1μs     0.61  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-     18.5±0.02μs       11.1±0.2μs     0.60  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-      14.2±0.3μs       8.43±0.8μs     0.59  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-     12.6±0.08μs       7.41±0.5μs     0.59  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-      33.2±0.2μs       19.5±0.6μs     0.59  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)
-     13.0±0.06μs       7.33±0.4μs     0.56  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-      28.3±0.1μs       11.1±0.3μs     0.39  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-     28.2±0.02μs       11.1±0.3μs     0.39  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-      28.2±0.1μs       11.0±0.3μs     0.39  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-     28.5±0.05μs       11.1±0.3μs     0.39  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-     77.7±0.04μs      4.52±0.07μs     0.06  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
-     77.6±0.01μs      4.47±0.04μs     0.06  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-      82.7±0.1μs      4.24±0.04μs     0.05  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
```
</details>

---

### Power little-endian (Power9/VSX3)

<details>
  <summary>CPU</summary>
  
```
Machine
Architecture:                    ppc64le
Byte Order:                      Little Endian
CPU(s):                          160
On-line CPU(s) list:             0-159
Thread(s) per core:              4
Core(s) per socket:              20
Socket(s):                       2
NUMA node(s):                    2
Model:                           2.2 (pvr 004e 1202)
Model name:                      POWER9, altivec supported
Frequency boost:                 enabled
CPU max MHz:                     3800.0000
CPU min MHz:                     2166.0000
L1d cache:                       1.3 MiB
L1i cache:                       1.3 MiB
L2 cache:                        10 MiB
L3 cache:                        200 MiB
NUMA node0 CPU(s):               0-79
NUMA node8 CPU(s):               80-159
```
</details>

<details>
  <summary>OS</summary>

```
gcc (GCC) 11.2.1 20210921
Ubuntu 20.04.3 LTS
```
</details>

#### Benchmark

<details>
  <summary>VSX3</summary>

```
        before          after
-      17.2±0.4μs       14.9±0.5μs     0.86  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-      11.9±0.2μs       9.93±0.3μs     0.83  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-      18.2±0.5μs       14.6±0.3μs     0.80  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-      12.6±0.4μs       9.72±0.3μs     0.77  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-      20.5±0.3μs       15.7±0.4μs     0.76  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-      18.6±0.2μs       14.0±0.1μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-     20.7±0.05μs       15.2±0.3μs     0.73  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-      30.0±0.2μs       21.7±0.2μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-      30.4±0.4μs       22.0±0.4μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-      12.8±0.1μs       9.29±0.4μs     0.72  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int16'>)
-      30.2±0.2μs       21.8±0.2μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-      35.0±0.3μs       25.0±0.6μs     0.72  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-      30.4±0.5μs       21.8±0.2μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-      10.3±0.3μs       6.96±0.2μs     0.68  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int8'>)
-        15.9±2μs       10.1±0.3μs     0.63  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
-        43.2±8μs      24.6±0.08μs     0.57  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)
-        28.2±6μs       14.3±0.7μs     0.51  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-      56.5±0.8μs       24.6±0.3μs     0.44  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-      57.0±0.3μs       22.1±0.5μs     0.39  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-      57.3±0.7μs       21.7±0.2μs     0.38  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
-      70.8±0.5μs       15.3±0.2μs     0.22  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
-      68.7±0.5μs       14.1±0.4μs     0.21  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float32'>)
-        112±30μs       14.1±0.5μs     0.13  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float32'>)
-        72.5±9μs       8.25±0.2μs     0.11  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
-      74.2±0.2μs       8.04±0.1μs     0.11  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-        75.2±1μs      7.78±0.07μs     0.10  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)

```
</details>

---

### AArch64

As I do not have access to an ARM processor, I didn't execute the benchmark to check performance. But I was able to test the NEON code I added in this PR on a small ARM processor.

cc: @mattip and @seiko2plus

rafaelcfsousa committed:
SIMD: Use universal intrinsics to implement comparison functions

mattip commented on numpy/core/src/umath/loops_comparison.dispatch.c.src:
```diff
@@ -0,0 +1,421 @@
+    for (; len >= vstep; len -= vstep, src += vstep, dst += vstep) {
+#if @len@ >= 8
+        npyv_@sfx@  b1 = npyv_load_@sfx@(src + npyv_nlanes_@sfx@ * 0);
+        npyv_b@len@ c1 = npyv_@VOP@_@sfx@(a, b1);
```
It seems only this path is followed? Are we missing tests?

rafaelcfsousa replied:
Yes, it was missing a test for the comparison functions. I just added the missing test in a new commit in this PR (https://github.com/numpy/numpy/pull/21483/commits/8da710018a0cf5ee7b35cc12f8af9dfa713789bd).

mattip commented on numpy/core/src/umath/loops_comparison.dispatch.c.src:
```diff
@@ -0,0 +1,421 @@
+    for (; len >= vstep; len -= vstep, src += vstep, dst += vstep) {
+        npyv_b8 b = npyv_cmpneq_u8(npyv_load_u8(src), vzero);
+        npyv_b8 c = npyv_@VOP@_u8(a, npyv_cvt_u8_b8(b));
+        npyv_store_u8(dst, npyv_and_u8(npyv_cvt_u8_b8(c), truemask));
```
This code is also unreached by coverage?

mattip commented on numpy/core/tests/test_simd.py:
```diff
@@ -156,6 +156,40 @@ def test_tobits(self):
             tobits = bin(self.tobits(vdata))
             assert tobits == bin(data_bits)
 
```
Yay, tests!

rafaelcfsousa replied:
This one only tests the new universal intrinsics.

mattip commented:
@seiko2plus thoughts? This significantly speeds up integer comparisons

rafaelcfsousa commented:
Just noticed that I also have to add a test for array OP scalar (and scalar OP array).

rafaelcfsousa commented:
I already have the required tests. The PR is ready to be reviewed again. Thanks :)
cc: @mattip and @seiko2plus

seiko2plus commented on numpy/core/src/umath/loops_comparison.dispatch.c.src:
```diff
@@ -0,0 +1,421 @@
+        // Whatever element in src != 0x0 is converted to 0xFF
+        npyv_b8 a = npyv_cmpneq_u8(npyv_load_u8(src1), vzero);
+        npyv_b8 b = npyv_cmpneq_u8(npyv_load_u8(src2), vzero);
+        npyv_b8 c = npyv_@VOP@_u8(npyv_cvt_u8_b8(a), npyv_cvt_u8_b8(b));
```
```suggestion
        npyv_b8 c = npyv_@VOP@_b8(a, b);
```
logical bitwise should be used instead, faster in general and its necessary for AVX512 to avoid the plenty of mask conversion.  e.g. `==` -> `and`, `!=` `xor`, `<` => `!a & b`(it would better to implement intrinsic for `andc` ), `<=` `!a | b` same here but for `orc`

rafaelcfsousa replied:
Done.

See below the mapping (cmp ==> logical bitwise)
`==` ==> `xnor`
`!=` ==>   `xor`
`<` ==> `~a & b  (andc)`
`<=` ==> `~a | b (orc)`
`>` ==> `~b & a  (reverse andc)`
`>=` ==> `~b | a (reverse orc)`

I introduced three new universal intrinsics, as listed below:
- `npyv_andc_b8`
- `npyv_orc_b8`
- `npyv_xnor_b8`

Well, as I am using only for `b8` I created a version only for this datatype. Should I create the same intrinsics for the other dtypes too?

See below the improvements compared to the first commit of this PR.

SSE
```
-     6.78±0.03μs      5.69±0.05μs     0.84  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-     6.57±0.01μs      5.10±0.01μs     0.78  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
```

AVX2
```
-     5.21±0.01μs      4.68±0.02μs     0.90  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
-     5.30±0.02μs      4.65±0.03μs     0.88  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
```


AVX512F
```
-      5.12±0.2μs       4.26±0.1μs     0.83  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-      5.26±0.2μs       4.34±0.1μs     0.83  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
-      5.49±0.1μs       4.45±0.1μs     0.81  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
```


AVX512BW
```
-      5.08±0.2μs       4.36±0.1μs     0.86  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
-      5.19±0.2μs       4.34±0.1μs     0.84  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-      5.45±0.1μs       4.46±0.1μs     0.82  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
```


VSX
```
-     6.69±0.01μs      5.91±0.02μs     0.88  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-     6.69±0.01μs      5.90±0.01μs     0.88  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
```

Thank you!

seiko2plus replied:
> Well, as I am using only for b8 I created a version only for this datatype. Should I create the same intrinsics for the other dtypes too?

not necessary, but we may need to add `andc` for u8 so we can use `cmpeq` instead of `cmpneq` when to comes BOOL data type.

rafaelcfsousa replied:
Done! I added `andc` for `u8`.

seiko2plus reviewed (changes requested):
Necessary improvements to the x86 architecture

rafaelcfsousa commented on numpy/core/src/common/simd/avx2/conversion.h:
```diff
@@ -58,6 +58,34 @@ NPY_FINLINE npyv_u32x2 npyv_expand_u32_u16(npyv_u16 data) {
+    __m256i p4 = _mm256_permute4x64_epi64(_mm256_packs_epi32(g, h), 216);
+    return npyv_pack_b8_b32(p1, p2, p3, p4);
+}
+
```
Done! Removing the unnecessary shuffles leads to a speedup of up to 11% compared to the initial solution of this PR (see below).

```
export NPY_DISABLE_CPU_FEATURES="AVX512F AVX512_SKX"
+     9.31±0.02μs      9.90±0.06μs     1.06  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-     17.4±0.03μs      16.5±0.05μs     0.95  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-     17.4±0.02μs      16.5±0.05μs     0.95  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-     14.8±0.07μs       14.0±0.1μs     0.95  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-        5.05±0μs      4.66±0.01μs     0.92  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint8'>)
-     17.4±0.04μs      16.0±0.07μs     0.92  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-     17.3±0.04μs      15.9±0.05μs     0.92  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-     5.80±0.01μs      5.18±0.03μs     0.89  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
```

rafaelcfsousa commented on numpy/core/src/common/simd/avx512/conversion.h:
```diff
@@ -90,6 +90,96 @@ NPY_FINLINE npyv_u32x2 npyv_expand_u32_u16(npyv_u16 data)
+#endif
+}
+
+#undef NPYV__PACK_LO_HI
```
I used your code to implement the function `npyv_pack_b8_b64` (AVX512F) with a few changes (see below):

```
    __m512i ai = npyv_cvt_u32_b32(a);
    __m512i bi = npyv_cvt_u32_b32(b);
    __m512i ci = npyv_cvt_u32_b32(c);
    __m512i di = npyv_cvt_u32_b32(d);
    __m512i ab = npyv512_packs_epi32(ai, bi);
    __m512i cd = npyv512_packs_epi32(ci, di);
+    __m512i abcd = npyv512_packs_epi16(ab, cd);  <-- added
+    return npyv512_shuffle_odd32(abcd);          <-- added
-    return return npyv_pack_b16(ab, cd);         <-- removed
```

and I noticed some slowdown comparing your solution vs mine:

```
AVX512F:
export NPY_DISABLE_CPU_FEATURES="AVX512_SKX"

+      8.98±0.3μs      10.4±0.09μs     1.16  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
+      8.99±0.4μs      10.4±0.07μs     1.15  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
+     9.73±0.05μs      11.2±0.08μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
+     9.78±0.07μs      11.3±0.09μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
+      9.83±0.1μs       11.3±0.1μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float32'>)
+     9.77±0.07μs      11.2±0.08μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
+      9.81±0.1μs       11.3±0.1μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float32'>)
+     9.79±0.05μs       11.2±0.2μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
+      9.11±0.4μs      10.4±0.08μs     1.14  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-     19.1±0.06μs      14.3±0.07μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-     19.1±0.09μs      14.3±0.07μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-     19.2±0.08μs      14.3±0.05μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-     19.2±0.06μs      14.3±0.09μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-     19.3±0.07μs      14.4±0.01μs     0.74  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-      19.3±0.5μs      14.3±0.05μs     0.74  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
```

What is happening (IMO) is that your code does some conversions that can be avoided. Keeping the data as __m256i and only converting it back to __m512i at the end of all computation seems to be better. For this case, in particular, I am keeping my code but with some small changes. If you want me to change it, let me know.


See below the improvements after changing the code and comparing it with the initial solution of this PR.

#### AVX512F:

```
export NPY_DISABLE_CPU_FEATURES="AVX512_SKX"

-      20.5±0.2μs       18.1±0.9μs     0.88  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)                                                                      
-      20.4±0.4μs         17.9±1μs     0.87  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)                                                                        
-        21.6±2μs         16.5±2μs     0.76  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)                                                                       
-     19.2±0.08μs      12.8±0.01μs     0.67  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)                                                                       
-     19.2±0.06μs      12.8±0.01μs     0.66  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)                                                                     
-     19.2±0.08μs      12.8±0.03μs     0.66  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)                                                                       
-     19.2±0.07μs      12.7±0.02μs     0.66  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)                                                                      
-      19.3±0.3μs      12.8±0.03μs     0.66  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)                                                                      
-      19.7±0.4μs      12.8±0.06μs     0.65  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>) 
``` 


#### AVX512BW:

```
unset NPY_DISABLE_CPU_FEATURES

-     5.42±0.09μs      5.15±0.09μs     0.95  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-     5.38±0.06μs      5.06±0.02μs     0.94  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-     5.41±0.04μs      5.08±0.03μs     0.94  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-     5.45±0.06μs      5.05±0.03μs     0.93  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
```

seiko2plus replied:
> What is happening (IMO) is that your code does some conversions that can be avoided.

I agree, there're extra shuffles in my version(AVX512F only), feel free to modified but keep using  `_mm512_kunpackd, _mm512_kunpackw, _mm512_kunpackb` to avoid unnecessary mask conversion.

rafaelcfsousa commented on numpy/core/src/umath/loops_comparison.dispatch.c.src:
```diff
@@ -0,0 +1,421 @@
+/*@targets
+ ** $maxopt baseline
+ ** sse2 sse41 avx2 avx512f avx512_skx
+ ** vsx3
```
Done, thanks!!

h-vetinari commented on numpy/core/tests/test_umath.py:
```diff
@@ -185,6 +185,82 @@ def __array_wrap__(self, arr, context):
 
 
 class TestComparisons:
+    @pytest.mark.parametrize('dtype', np.sctypes['uint'] + np.sctypes['int'] +
+                             np.sctypes['float'] + [np.bool_])
+    def test_comparison_functions(self, dtype):
```
```suggestion
    @pytest.mark.parametrize('dtype', np.sctypes['uint'] + np.sctypes['int'] +
                             np.sctypes['float'] + [np.bool_])
    @pytest.mark.parametrize('py_comp,np_comp', [
        (operator.lt, np.less),
        (operator.le, np.less_equal)
        (operator.gt, np.greater)
        (operator.ge, np.greater_equal)
        (operator.eq, np.equal)
        (operator.ne, np.not_equal)
    ])
    def test_comparison_functions(self, dtype, py_comp, np_comp):
```

rafaelcfsousa replied:
Done! Thank you! I tried this mapping but I didn't know I could use `operator.*`.

h-vetinari commented on numpy/core/tests/test_umath.py:
```diff
@@ -185,6 +185,82 @@ def __array_wrap__(self, arr, context):
+        a_lst = a.tolist()
+        b_lst = b.tolist()
+
+        # (Binary) Comparison (x1=array, x2=array)
+        lt_b = np.less(a, b)
+        le_b = np.less_equal(a, b)
+        gt_b = np.greater(a, b)
+        ge_b = np.greater_equal(a, b)
+        eq_b = np.equal(a, b)
+        ne_b = np.not_equal(a, b)
+        lt_b_lst = [x < y for x, y in zip(a_lst, b_lst)]
+        le_b_lst = [x <= y for x, y in zip(a_lst, b_lst)]
+        gt_b_lst = [x > y for x, y in zip(a_lst, b_lst)]
+        ge_b_lst = [x >= y for x, y in zip(a_lst, b_lst)]
+        eq_b_lst = [x == y for x, y in zip(a_lst, b_lst)]
+        ne_b_lst = [x != y for x, y in zip(a_lst, b_lst)]
```
and then...

```suggestion
        # (Binary) Comparison (x1=array, x2=array)
        comp_b = np_comp(a, b)
        comp_b_list = [py_comp(x, y) for x, y in zip(a_lst, b_lst)]
```

(likewise for `_s1` & `_s2`)

h-vetinari commented on numpy/core/tests/test_umath.py:
```diff
@@ -185,6 +185,82 @@ def __array_wrap__(self, arr, context):
+        eq_s2_lst = [x == scalar for x in a_lst]
+        ne_s2_lst = [x != scalar for x in a_lst]
+
+        # Compare comparison functions (Python vs NumPy) using native Python
+        def compare(lt, le, gt, ge, eq, ne, lt_lst, le_lst, gt_lst, ge_lst,
+                    eq_lst, ne_lst):
+            assert_(lt.tolist() == lt_lst, "Comparison function check (lt)")
+            assert_(le.tolist() == le_lst, "Comparison function check (le)")
+            assert_(gt.tolist() == gt_lst, "Comparison function check (gt)")
+            assert_(ge.tolist() == ge_lst, "Comparison function check (ge)")
+            assert_(eq.tolist() == eq_lst, "Comparison function check (eq)")
+            assert_(ne.tolist() == ne_lst, "Comparison function check (ne)")
+
+        # Sequence: Binary, Scalar1 and Scalar2
+        compare(lt_b, le_b, gt_b, ge_b, eq_b, ne_b, lt_b_lst, le_b_lst,
+                gt_b_lst, ge_b_lst, eq_b_lst, ne_b_lst)
+        compare(lt_s1, le_s1, gt_s1, ge_s1, eq_s1, ne_s1, lt_s1_lst, le_s1_lst,
+                gt_s1_lst, ge_s1_lst, eq_s1_lst, ne_s1_lst)
+        compare(lt_s2, le_s2, gt_s2, ge_s2, eq_s2, ne_s2, lt_s2_lst, le_s2_lst,
+                gt_s2_lst, ge_s2_lst, eq_s2_lst, ne_s2_lst)
```
and then

```suggestion
        assert_(comp_b.tolist() == comp_b_list, f"Failed comparision ({py_comp.__name__})")
        assert_(comp_s1.tolist() == comp_s1_list, f"Failed comparision ({py_comp.__name__})")
        assert_(comp_s2.tolist() == comp_s2_list, f"Failed comparision ({py_comp.__name__})")
```

h-vetinari reviewed (commented):
I feel this test would lend itself very well to parametrizing across the comparison too, examples below.

rafaelcfsousa committed:
TST: Rewrite the test that checks the universal SIMD intrinsic (pack)

rafaelcfsousa committed:
SIMD, ENH: Use logical bitwise to implement comparison functions (bool_)

rafaelcfsousa commented:
We have some errors in the CI but I am not sure if they are related to the latest changes.

seiko2plus commented on numpy/core/src/common/simd/avx2/operators.h:
```diff
@@ -114,6 +114,11 @@ NPY_FINLINE __m256i npyv_shr_s64(__m256i a, int c)
 
+// ANDC, ORC and XNOR
+#define npyv_andc_b8(A, B) _mm256_andnot_si256(A, B)
+#define npyv_orc_b8(A, B) npyv_or_b8(npyv_not_b8(A), B)
+#define npyv_xnor_b8(A, B) npyv_not_b8(npyv_xor_b8(A, B))
```
`_cmplt_epi8` and `_cmpeq_epi8` can be used for AVX2 and SSE since boolean values are guaranteed to be 0 or 0xFF, also because {8,16-32}-bit comparison take almost 0.5 latency.

rafaelcfsousa replied:
Done! I applied the suggested changes above for `xnor`, but not for `orc` since it would require a `less equal` intrinsic (which is not available neither for SSE nor AVX2).

seiko2plus commented on numpy/core/tests/test_simd.py:
```diff
@@ -156,6 +156,79 @@ def test_tobits(self):
+        if self.sfx not in ("b8"):
+            return
+        andc_simd = getattr(self.npyv, f"andc_b8")
+        # create the vectors
+        data = self._data()
+        rdata = self._data(reverse=True)
+        vdata = self._load_b(data)
+        vrdata = self._load_b(rdata)
+        # check andc
+        sandc = [(~x & y) & 0xFF for x, y in zip(rdata, data)]
+        vandc = andc_simd(vrdata, vdata)
+        assert sandc == vandc
+
+    def test_orc(self):
+        if self.sfx not in ("b8"):
+            return
+        orc_simd = getattr(self.npyv, f"orc_b8")
+        # create the vectors
+        data = self._data()
+        rdata = self._data(reverse=True)
+        vdata = self._load_b(data)
+        vrdata = self._load_b(rdata)
+        # check orc
+        sorc = [(~x | y) & 0xFF for x, y in zip(rdata, data)]
+        vorc = orc_simd(vrdata, vdata)
+        assert sorc == vorc
+
+    def test_xnor(self):
+        if self.sfx not in ("b8"):
+            return
+        xnor_simd = getattr(self.npyv, f"xnor_b8")
+        # create the vectors
+        data = self._data()
+        rdata = self._data(reverse=True)
+        vdata = self._load_b(data)
+        vrdata = self._load_b(rdata)
+        # check orc
+        sxnor = [~(x ^ y) & 0xFF for x, y in zip(rdata, data)]
+        vxnor = xnor_simd(vrdata, vdata)
+        assert sxnor == vxnor
```
```suggestion
        if self.sfx not in ("b8"):
            return
        # check andc
        sandc = [(~x & y) & 0xFF for x, y in zip(rdata, data)]
        vandc = andc_simd(vrdata, vdata)
        assert sandc == vandc
        # check orc
        sorc = [(~x | y) & 0xFF for x, y in zip(rdata, data)]
        vorc = orc_simd(vrdata, vdata)
        assert sorc == vorc
        # check xnor
        sxnor = [~(x ^ y) & 0xFF for x, y in zip(rdata, data)]
        vxnor = xnor_simd(vrdata, vdata)
        assert sxnor == vxnor
```
you can combine these cases with the above section `test_operators_logical` in order to reduce the amount of code

rafaelcfsousa replied:
Done! I reduced the code above and also added a test for andc (u8).

seiko2plus reviewed (changes requested):
Thank you, we're almost done here just few changes more.

rafaelcfsousa committed:
SIMD, ENH: Add universal intrinsic andc8 and use it to remove ifneq

This commit also applies some techniques to reduce the size of the binary generated from the source loops_comparison.dispatch.c.src

rafaelcfsousa committed:
TST: Add test for andc (u8)

This commit also rewrite the tests andc, orc and xnor

rafaelcfsousa commented on numpy/core/src/umath/loops_comparison.dispatch.c.src:
```diff
@@ -0,0 +1,434 @@
+#else
+        npyv_b8 c = npyv_@VOP@_b8(b, a);
+#endif
+        npyv_store_u8(dst, npyv_and_u8(npyv_cvt_u8_b8(c), truemask));
```
Done! I used npyv_cmpeq_u8 and npyv_andc_u8 on the places you pointed out above.

rafaelcfsousa commented on numpy/core/src/common/simd/avx512/conversion.h:
```diff
@@ -90,6 +90,48 @@ NPY_FINLINE npyv_u32x2 npyv_expand_u32_u16(npyv_u16 data)
+npyv_pack_b8_b32(npyv_b32 a, npyv_b32 b, npyv_b32 c, npyv_b32 d) {
+#ifdef NPY_HAVE_AVX512BW
+    __mmask32 ab = (__mmask64)_mm512_kunpackw((__mmask32)b, (__mmask32)a);
+    __mmask32 cd = (__mmask64)_mm512_kunpackw((__mmask32)d, (__mmask32)c);
```
Done! The unnecessary casts were removed.

rafaelcfsousa commented on numpy/core/src/common/simd/avx2/operators.h:
```diff
@@ -114,6 +114,11 @@ NPY_FINLINE __m256i npyv_shr_s64(__m256i a, int c)
 #define npyv_not_b64 npyv_not_u8
 
+// ANDC, ORC and XNOR
+#define npyv_andc_b8(A, B) _mm256_andnot_si256(A, B)
```
Done! I applied the rules you show above. I have a table below in my last comment that shows the binary size after this optimization.

rafaelcfsousa commented on numpy/core/src/umath/loops_comparison.dispatch.c.src:
```diff
@@ -0,0 +1,434 @@
+    npy_clear_floatstatus_barrier((char*)dimensions);
+}
+/**end repeat1**/
+/**end repeat**/
```
Done! I applied the rules you show above. I have a table below in my last comment that shows the binary size after this optimization.

rafaelcfsousa commented:
See below the results updated:


### Benchmark/Performance:

<details>
  <summary>SSE</summary>

```
export NPY_DISABLE_CPU_FEATURES="AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_SKX"

       before           after         ratio
     [ae8b9ce9]       [04f68033]
     <main>           <simd_comparison>
+     8.24±0.07μs      8.92±0.02μs     1.08  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint16'>)
+     5.40±0.04μs      5.76±0.02μs     1.07  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint8'>)
+     5.37±0.02μs      5.70±0.07μs     1.06  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint8'>)
-      17.8±0.1μs       16.0±0.2μs     0.90  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
-     5.84±0.02μs      5.11±0.01μs     0.87  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint8'>)
-      50.8±0.1μs       43.6±0.2μs     0.86  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-      8.92±0.1μs      7.61±0.03μs     0.85  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-     8.72±0.09μs      7.37±0.04μs     0.85  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-      50.4±0.1μs       42.5±0.1μs     0.84  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-     7.70±0.07μs       6.48±0.2μs     0.84  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
-        30.0±2μs       25.1±0.6μs     0.84  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-     17.9±0.03μs       14.8±0.1μs     0.83  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-      50.4±0.3μs       40.0±0.4μs     0.79  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-     50.5±0.05μs      39.0±0.04μs     0.77  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-      50.4±0.1μs       38.1±0.1μs     0.76  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-     20.1±0.04μs      14.8±0.05μs     0.73  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-     9.81±0.02μs       6.83±0.2μs     0.70  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-     20.2±0.09μs       13.0±0.1μs     0.64  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-      19.0±0.1μs      11.6±0.07μs     0.61  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-     18.8±0.04μs      11.3±0.02μs     0.60  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-     18.7±0.03μs      9.46±0.08μs     0.50  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-     77.5±0.05μs      5.76±0.03μs     0.07  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-     77.6±0.02μs      5.40±0.04μs     0.07  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
-     76.5±0.05μs      5.32±0.02μs     0.07  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
```
</details>

<details>
  <summary>AVX2</summary>

```
export NPY_DISABLE_CPU_FEATURES="AVX512F AVX512CD AVX512_SKX"

-      13.5±0.3μs      12.7±0.03μs     0.94  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-      7.28±0.1μs       6.65±0.1μs     0.91  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-      6.72±0.1μs      6.11±0.07μs     0.91  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
-      7.38±0.2μs      6.67±0.05μs     0.90  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-      6.77±0.1μs      6.04±0.06μs     0.89  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-      14.5±0.2μs      12.8±0.05μs     0.89  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
-      11.0±0.2μs      9.08±0.04μs     0.82  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-      11.1±0.2μs      9.04±0.05μs     0.81  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-     12.0±0.04μs      9.64±0.04μs     0.81  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-      12.1±0.2μs      9.67±0.03μs     0.80  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-      17.8±0.1μs       14.1±0.2μs     0.80  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
-     17.9±0.08μs      13.9±0.08μs     0.78  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-      28.5±0.5μs       22.2±0.2μs     0.78  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-      29.3±0.7μs       21.7±0.4μs     0.74  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-      32.4±0.8μs       23.8±0.7μs     0.74  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)
-     27.8±0.07μs       16.5±0.2μs     0.60  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-     27.9±0.07μs      16.5±0.04μs     0.59  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-     27.8±0.07μs      16.0±0.04μs     0.57  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-     27.8±0.08μs      16.0±0.03μs     0.57  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-      76.2±0.1μs         5.10±0μs     0.07  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
-      77.2±0.2μs      4.69±0.04μs     0.06  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
-      77.5±0.1μs      4.69±0.01μs     0.06  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
```
</details>

<details>
  <summary>AVX512F</summary>

```
export NPY_DISABLE_CPU_FEATURES="AVX512CD AVX512_SKX"

-      6.79±0.1μs       6.37±0.2μs     0.94  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
-      6.84±0.1μs       6.33±0.2μs     0.92  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-      7.30±0.1μs       6.71±0.3μs     0.92  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-      7.42±0.2μs       6.74±0.3μs     0.91  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-     4.54±0.01μs      4.08±0.09μs     0.90  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int8'>)
-     4.57±0.03μs      4.05±0.02μs     0.88  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int8'>)
-     4.71±0.03μs      4.16±0.09μs     0.88  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint8'>)
-     4.58±0.04μs      4.00±0.04μs     0.87  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int8'>)
-     4.72±0.03μs      4.08±0.09μs     0.87  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint8'>)
-     4.60±0.01μs       3.91±0.1μs     0.85  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint8'>)
-     11.7±0.04μs      9.72±0.03μs     0.83  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-     11.6±0.05μs      9.70±0.01μs     0.83  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-     8.14±0.06μs      6.55±0.04μs     0.80  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint16'>)
-     8.07±0.04μs      6.47±0.06μs     0.80  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int16'>)
-     12.6±0.03μs      9.73±0.02μs     0.77  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-     12.6±0.03μs       9.72±0.1μs     0.77  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-      18.0±0.2μs      12.9±0.03μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
-     18.3±0.08μs      12.8±0.04μs     0.70  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-     13.7±0.03μs      9.00±0.08μs     0.66  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-      29.0±0.3μs         18.9±1μs     0.65  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-     14.5±0.05μs      8.90±0.03μs     0.61  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-      14.6±0.2μs      8.93±0.06μs     0.61  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
-      32.1±0.3μs       18.1±0.8μs     0.56  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-      34.2±0.2μs       18.1±0.9μs     0.53  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)
-     27.9±0.03μs       12.9±0.3μs     0.46  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-     27.9±0.07μs      12.8±0.04μs     0.46  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-      28.3±0.2μs      12.8±0.08μs     0.45  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-     28.2±0.07μs      12.8±0.02μs     0.45  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-     76.4±0.02μs      4.12±0.06μs     0.05  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
-     77.4±0.03μs      4.06±0.04μs     0.05  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-     77.4±0.03μs      4.04±0.05μs     0.05  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
```
</details>

<details>
  <summary>AVX512BW</summary>

```
unset NPY_DISABLE_CPU_FEATURES

       before           after         ratio
     [ae8b9ce9]       [04f68033]
     <main>           <simd_comparison>
-     4.49±0.01μs      4.27±0.03μs     0.95  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int8'>)
-        4.49±0μs      4.21±0.05μs     0.94  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int8'>)
-     4.52±0.01μs      4.23±0.05μs     0.94  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint8'>)
-     4.56±0.04μs      4.24±0.05μs     0.93  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint8'>)
-     4.56±0.01μs      4.01±0.02μs     0.88  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint8'>)
-     4.54±0.02μs      3.96±0.02μs     0.87  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int8'>)
-      9.41±0.1μs      7.08±0.02μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float32'>)
-      13.0±0.2μs      9.71±0.05μs     0.75  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-     6.91±0.04μs       5.06±0.2μs     0.73  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
-     6.82±0.09μs       4.96±0.1μs     0.73  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-     9.67±0.02μs      6.98±0.03μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float32'>)
-      28.1±0.1μs       20.3±0.4μs     0.72  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-      13.5±0.3μs      9.69±0.06μs     0.72  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-      28.8±0.2μs       20.3±0.8μs     0.71  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-     14.6±0.09μs      9.71±0.02μs     0.67  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
-     7.53±0.04μs       5.00±0.1μs     0.66  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-     8.03±0.02μs       5.26±0.4μs     0.66  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int16'>)
-     7.60±0.04μs       4.97±0.2μs     0.65  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-      11.0±0.1μs      6.94±0.04μs     0.63  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-     8.13±0.02μs       5.09±0.5μs     0.63  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint16'>)
-     11.1±0.02μs      6.95±0.02μs     0.62  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-      17.6±0.2μs      10.8±0.03μs     0.62  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
-     18.2±0.08μs      11.0±0.05μs     0.60  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-     31.7±0.07μs       19.1±0.4μs     0.60  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)
-     12.1±0.05μs      7.01±0.02μs     0.58  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-     12.2±0.08μs      7.00±0.02μs     0.58  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-     27.8±0.05μs      10.9±0.03μs     0.39  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-     27.8±0.04μs      10.9±0.08μs     0.39  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-     27.8±0.04μs      10.8±0.04μs     0.39  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-     27.8±0.03μs      10.8±0.03μs     0.39  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-     77.3±0.05μs      4.17±0.06μs     0.05  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-      77.4±0.1μs      4.16±0.03μs     0.05  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>)
-     76.4±0.02μs      3.97±0.01μs     0.05  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
```
</details>

<details>
  <summary>VSX3</summary>

```
-      11.4±0.8μs       8.90±0.9μs     0.78  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-        16.4±2μs         12.7±1μs     0.78  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
-        28.2±2μs         21.5±1μs     0.76  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-        28.0±2μs         20.3±1μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-        16.8±2μs         12.1±1μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
-        16.8±2μs         12.1±1μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
-        17.1±4μs         12.3±1μs     0.72  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
-        19.6±6μs         14.0±1μs     0.71  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
-        32.6±2μs         23.3±2μs     0.71  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)
-      11.9±0.8μs       8.47±0.7μs     0.71  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int16'>)
-        28.4±2μs         20.2±1μs     0.71  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-        19.6±2μs       13.9±0.9μs     0.71  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-        15.6±5μs       8.71±0.9μs     0.56  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint16'>)
-       42.6±10μs         23.6±2μs     0.55  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)
-      50.3±0.5μs         22.9±1μs     0.46  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)
-       19.6±10μs       8.73±0.9μs     0.44  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-        54.2±4μs         20.3±2μs     0.37  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
-        55.3±5μs         20.3±1μs     0.37  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-        65.2±5μs      13.2±0.07μs     0.20  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
-        65.0±4μs         13.1±1μs     0.20  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float32'>)
-        64.9±5μs         12.7±1μs     0.20  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float32'>)
-        67.8±5μs       7.34±0.6μs     0.11  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.bool_'>)
-        70.3±6μs       6.30±0.5μs     0.09  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.bool_'>)
-       75.9±70μs         6.66±2μs     0.09  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.bool_'>
```
</details>


### Binary size:

|File | main | simd | simd+binaryopt|
|---------------------------------|-----|------|------|
|_multiarray_umath.cpython-39-x86_64-linux-gnu.so | 5138K | 5603K | 5407K|
|_multiarray_umath.cpython-39-powerpc64le-linux-gnu.so| 30435K | 31712K | 30477K |

seiko2plus commented on numpy/core/src/umath/loops_comparison.dispatch.c.src:
```diff
@@ -0,0 +1,451 @@
+/*@targets
+ ** $maxopt baseline
+ ** sse2 sse41 avx2 avx512f avx512_skx
```
```suggestion
 ** sse2 sse42 avx2 avx512f avx512_skx
```
targets `sse41` just for `_mm_cmpeq_epi64` is kinda excess regards to cost of binary size, maybe `sse42` would be better choice at least it comes with `_mm_cmpgt_epi64`.

rafaelcfsousa replied:
Done!

seiko2plus reviewed (changes requested):
Well done!, just one thing left.

seiko2plus commented on benchmarks/benchmarks/bench_ufunc.py:
```diff
@@ -170,8 +170,25 @@ def time_divide_scalar2(self, dtype):
+        self.s = np.ones(1, dtype=dtype)
+
+    def time_less_than_binary(self, dtype):
+        (self.x < self.y)
+
+    def time_less_than_scalar1(self, dtype):
+        (self.s < self.x)
+
     def time_less_than_scalar2(self, dtype):
-        (self.d < 1)
+        (self.x < self.s)
```
@mattip, benchmarks here doesn't cover the rest of operations, not sure if its necessary to cover them or its better to save a room for the upcoming benchmarks since its the benchmark process start to became pretty slow.

mattip replied:
+1 for minimal effective benchmarks. I don't think there is a need to repeatedly hit the same code multiple times

seiko2plus commented:
@rafaelcfsousa, its better to strip the debugging symbols before reporting the binary size. also I think we need a release note.

mattip commented:
> I think we need a release note.

On the one hand, there are hints in ['numpy/doc/release/upcoming_changes'](https://github.com/numpy/numpy/tree/main/doc/release/upcoming_changes). On the other, I see we are not very diligent [about adding release notes](https://github.com/numpy/numpy/pulls?q=is%3Apr+is%3Aclosed+label%3A%22component%3A+SIMD%22+label%3A%2256+-+Needs+Release+Note.%22) to other PRs for SIMD performance.

rafaelcfsousa committed:
DOC: Add a release note for the comparison functions

The PR #21483 improves the execution time of the comparison functions by using universal intrinsics

rafaelcfsousa commented:
> @rafaelcfsousa, its better to strip the debugging symbols before reporting the binary size. also I think we need a release note.

@seiko2plus :
For some reason, the compiler I was using on Power9/VSX3 was adding the debug flag `-g` even without specifying the use of the flag `--debug` to build NumPy. After using a different compiler, now I see the following for Power/VSX3:

|File | main | simd | simd+binaryopt|
|---------------------------------|-----|------|------|
|_multiarray_umath.cpython-39-powerpc64le-linux-gnu.so| 6661K | 6666K | 6603K |

rafaelcfsousa commented:
@seiko2plus and @mattip :
I added a release note.

seiko2plus reviewed (approved):
LGTM, thank you!

mattip commented:
Thanks @rafaelcfsousa and thanks @seiko2plus for the careful review
